### ISSUE TITLE
Question: Autocomplete Improvements

### ISSUE THREAD

**Issue by jfelchner** [neovim/neovim#396]
The autocomplete in Vim is abysmal when compared with IDEs.  Even the plugins which try to improve it such as NeoComplCache and SuperTab can't do what they need to do in order to be on par with Emacs or even SublimeText.

So my question, will neovim's refactorings allow either vim itself or plugin makers to be able to create a reasonable autocomplete system?

**Comment by dontdieych** [comment]
@Valloric Thanks for insightful comment. :+1:

**Comment by ChrisPenner** [comment]
I'm in favour of using the model suggested by @Valloric, one request I have is that Fuzzy completion be supported in this model from the beginning; including case-insensitive fuzzy-complete. 
Fuzzy completion helps substantially when typing constants like WINDOW_SIZE_HORIZONTAL and WINDOW_SIZE_VERTICAL which are hard to type and only differ at the end, and could also be used when specifying file paths etc. It's gaining popularity pretty quickly lately.

**Comment by jfelchner** [comment]
@ChrisPenner I agree.

**Comment by jfelchner** [comment]
Say what you will about Microsoft, but IntelliSense is freaking amazing.

Check out the video on https://code.visualstudio.com for a look at how they use it with JavaScript.  Very impressive.

This is the kind of thing that NeoVim should enable other developers to build plugins for.

![image](https://cloud.githubusercontent.com/assets/285582/7672303/2077f89c-fcb3-11e4-9cd4-d9b2e02f8993.png)

![image](https://cloud.githubusercontent.com/assets/285582/7672305/2802718c-fcb3-11e4-8a0f-bec6bf6abc00.png)

**Comment by jfelchner** [comment]
The other best in breed in my opinion is IntelliJ:

https://www.jetbrains.com/idea/features/code_completion.html

**Comment by renstrom** [comment]
As far as other terminal applications go, I think [bpython](http://www.bpython-interpreter.org) does this pretty nice:

![screen shot 2015-05-18 at 7 45 26](https://cloud.githubusercontent.com/assets/177685/7674933/e79daee8-fd31-11e4-903d-8a475f864c4f.png)

**Comment by jfelchner** [comment]
@renstrom very nice!

**Comment by gaving** [comment]
Throwing http://blog.atom.io/2015/05/15/new-autocomplete.html into the mix, hands down the nicest autocompletion I've seen yet.

![atom](https://cloud.githubusercontent.com/assets/69169/7659203/1ff15662-faf5-11e4-9404-086812a0d019.png)

**Comment by jfelchner** [comment]
@gaving excellent!

Keep the examples coming. I know that if we see what everyone else is doing we can take those ideas, improve on them, and make vim more amazing than ever. 😄

**Comment by a-b** [comment]
While vim amazingness coming is there any working autocompletion for neovim exists?

**Comment by gglanzani** [comment]
@a-b Yes, https://github.com/Shougo/deoplete.nvim

I'm using it on a daily basis and, although it's still rough around the edges, it's good.

**Comment by 7belrhalmia** [comment]
@tarruda  there is any improvement in neovim autocompletion

**Comment by jfelchner** [comment]
@7belrhalmia lots of behind the scenes upgrades which will eventually help but nothing will happen with this until probably version 1.0

**Comment by 7belrhalmia** [comment]
@jfelchner  but you think that autocomletion will be improved and will be like ide style like insert all function prototype or a popup to show function paramettre

**Comment by 7belrhalmia** [comment]
@jfelchner  and how long we have to wait until that will hapen

**Comment by jfelchner** [comment]
@7belrhalmia I wouldn't even begin to fathom a guess. It's gonna be a while. But they finally have "mini milestones" before 1.0 so we should see some visible forward progress soon.

**Comment by cHoco** [comment]
I love vim but I'm fed up with the way the complete popup + preview behaves, so I've been toying with a new design for it. Here's a couple of screenshots of how it would look. 
Results list:
![mockup1](https://cloud.githubusercontent.com/assets/85527/11729239/80560122-9f8e-11e5-831a-785a4368d003.png)
Selected element:
![mockup2](https://cloud.githubusercontent.com/assets/85527/11729238/805118ec-9f8e-11e5-8106-8f977ef99229.png)

Instead of opening a preview window, additional infos would get displayed under the list of results. The current patch I'm using is a rough PoC and since implementing it right would require rewriting a good chunk of `popupmenu.c` I decided to post here and get some feedback.
The patch would:
- render the `kind` of the suggestion first, and styles it based on the kind value.
- render the rest of the popup menu as before
- render extra info under the results list with another style attribute

This change wouldn't modify the current API, just change the design of the popup, so if collaborators prefer I can open a new issue instead of cluttering this one that seems more focused on improving the API itself.

**Comment by scampi** [comment]
@cHoco looks nice!

**Comment by jfelchner** [comment]
@cHoco beautiful!  Tell ya what, go ahead and open new issue and the mods can add your issue to the Autocomplete milestone.

**Comment by Shougo** [comment]
@cHoco It is really nice!  Can you send the Pull Request?

**Comment by PixelPirate** [comment]
@cHoco This looks really nice! I would love to use this!

**Comment by gavsim** [comment]
@cHoco Looks great 👍🏻

**Comment by Valloric** [comment]
@cHoco That looks amazing! 

I've been saying for years that Vim needs an API that plugins can use to create sub-windows. This could then be used to build whatever the plugins want in terms of code-completion display.

For instance, the screenshot you're showing puts the detailed info window at the bottom of the completion list. Personally, I think that putting it to the right of the completion list is much better, for various reasons. But if the window is being drawn by Vim/Neovim, then I as a plugin author have no choice but to use what you've provided.

So the whole "specific API in Vim for showing code-completions" is IMO too limiting. Better to provide generic support for drawing & placing custom sub-windows (and their styling) in plugins. This supports _so many_ use-cases, of which better code-completion is but one.

**Comment by danihodovic** [comment]
@cHoco Nice!

I agree with @Valloric. This is a bigger issue than just the completion popup, being able to draw custom windows opens up a world of new use cases for plugin creators.

**Comment by justinmk** [comment]
I've mentioned in some places that allowing buffers to "float" at arbitrary coordinates (while improving the buffer API) may be the best way to allow plugins to create arbitrary popups/menus (even "widgets", if you will). Though I did not create an issue for this concept.

**Comment by sanwv** [comment]
Nice! api to create/control sub-windows (or float widgets) also great!

**Comment by Valloric** [comment]
@justinmk Making the windows real buffers creates problems though. Suddenly what should be a display-only window can be navigated to (and possibly changed) by the user or (accidentally) by a separate plugin. 

As a plugin author, I want a window strictly under my control without worrying about a stray plugin or the user messing with it by accident. (The user should of course have the power to kill any such windows I create. The user is of course free to not use my plugin too.)

**Comment by wsdjeg** [comment]
@cHoco 
great work ! can you open an PR
@artur-shaik
MAYBE YOU SHOILD KNOW THIS ,IF IT IS MERGED,IT WILL BE EASY FOR YOU TO CREATE JAVADOC FOR EVERY METHOD OR FIELDS

**Comment by justinmk** [comment]
> As a plugin author, I want a window strictly under my control without worrying about a stray plugin or the user messing with it by accident

@Valloric Certainly the design is up for discussion. New facilities for such control may need to be invented, but I'll note that some clumsy ones exist already: `:set eventignore=all` and `:noautocmd`...

Main goal of re-using buffers for such purpose is to (1) inform new buffer-API design in service of such use-cases, and (2) to benefit users and nvim itself by re-using existing concepts.

**Comment by jfelchner** [comment]
@justinmk @Valloric either way, the fact that these discussions are being had, and that we all know improvements need to be made (whatever shape they take) is what makes Neovim amazing.

The great communication between plugin authors and Neovim core is outstanding.  After all, I think the best philosophy for Neovim is to understand what is the _underlying_ problem that plugin authors need solved and build that.  Because once those problems are solved, plugin authors can play around with different higher-level abstractions to find out what works best.  And Neovim core remains svelte and compact (well, as much as possible while still maintaining 50 years of backward compatibility :grinning:)

**Comment by xcambar** [comment]
@cHoco Great work! Off-topic: You have an amazing font and theme. Care to share?

**Comment by carlitux** [comment]
@cHoco that looks nice!!!

**Comment by MichaelMackus** [comment]
Nice, @cHoco !! That looks really cool.

Beyond just the styling of the popup, I definitely agree that vim's autocompletion engine could use some more love. The API is terrible in my little experience hacking things in the attempt to "combine" auto completion engines. I'd love some sort of completion frontend integrated into neovim, where plugin authors could just hook in their backend code (independent of language).

**Comment by PixelPirate** [comment]
Is there actuallly any progress on the new UI or a branch to test it / help?

**Comment by mhartington** [comment]
@cHoco not to be a pest, but this is awesome. Would love to see this added! Please make a PR or post a gist of the changes

**Comment by justinmk** [comment]
@mhartington  [#4432](https://github.com/neovim/neovim/pull/4432#issuecomment-240901890) shows a working POC by @romgrk using the new externalized popupmenu API.

**Comment by cHoco** [comment]
I'm sorry for never posting an update, but that's because there was nothing concrete to update about :(
The patch I used in the screenshot was basically just a hack not really ready for use, and the extra info below was actually completely hardcoded (it wasn't generated dynamically for each element).
When I started thinking about how to implement this change the right way, more than a single problem became evident:

1. in the patch, I had hardcoded the check for either 'f' 'c' 'm' etc... but these letters aren't actually standard and different completion plugins/omnicompletetion functions implement different letters etc.. so a first step would be to define all the possible values there (the documentation talks about some of them, but to correctly implement this all should be defined). Then once that is done, since I think everyone would love to costumize them, we'd need a to define for each of these letter types an highlight (more options wohooooo). And don't forget the highlight styles for the other components of the popupmenu (even more options)
2. the design above is able to keep the current behaviour of vim to align the completed word with the suggestion only because I use line numbers, but if one doesn't then the completion text wouldn't be aligned. That's when I thought that maybe people at that point would prefer to write the formatting themselves to better suit their needs, maybe with the same syntax of the status line. My idea was then to reuse the parsing function for status line to add and option to decide how to format the popupmenu (an expression like `statusline`), but I never had time to get familiar with that code to check if it was really feasible. This approach would also need a new option and not everyone likes them.
3. how to implement the the extra info additional menu? the issue is that drawing custom component like that in vim is rough, and it would also mean reimplement for each one a lot of the same logic. For example the popupmenu needs to check the text orientation, how much space it has to render and a bunch of other conditions to render correctly. The new extra_info popup would basically have to reimplement all of them! Plus I'd need to implement an algorithm to format the text to render inside that box and that would be another good chunk of conditions. So instead of all this reimplementation crap, wouldn't it be better to have an api to draw `windows` like the popupmenu to make developing these easier and also give access to the plugins? But again, I didn't really have time to get comfortable with the code and figure out a decent way to do this.
4. would it just be better to change to completion api altogether instead of keeping the current one (with the division defined for the popupmenu?  

I probably forgot other issues I though about when trying to come up with a decent design for these features, but these were the main ones. To continue working on this change I wanted to post about these issues and discuss the possible best solutions. But I really didn't have time then and when finally I could I had lost interest. Then the refactoring in https://github.com/neovim/neovim/pull/4432 came along and I was not sure how it would affect this, but now that it is merged, if I understand it right, it is geared towards external GUIs.
To conclude, before going forwards I think it would be better to hear the developers opinions about the issues above.

**Comment by Shougo** [comment]
I have found gonvim.
It implements the auto completion window.
It is awesome!

https://github.com/dzhou121/gonvim

**Comment by justinmk** [comment]
## Resolution/Summary

- > allowing buffers to "float" at arbitrary coordinates (while improving the buffer API) may be the best way to allow plugins to create arbitrary popups/menus (even "widgets", if you will).
  - Floating windows were implemented in 0.4.x by @bfredl in https://github.com/neovim/neovim/pull/6619 .
- Externalized popupmenu was implemented in #4432, and is used by several UI clients including Oni, Oni2, gNvim, nvim-gtk, veonim, neovim-qt, etc.
- Further structure around floating-windows is tracked by https://github.com/neovim/neovim/issues/9663 .


---
_Locked to keep the summary visible. You can always [chat](https://neovim.io/#chat) or open a ticket if you have new information/topics to discuss._
